Sources/GRPCHTTP2Core/Client/Connection/LoadBalancers/Subchannel.swift

```diff
@@ -0,0 +1,591 @@
+      }
+
+      var reason: Reason
+      var connection: Connection
+
+      init(from state: Connecting, reason: Reason) {
+        self.connection = state.connection
+        self.reason = reason
+      }
+
+      init(from state: Connected, reason: Reason) {
+        self.connection = state.connection
+        self.reason = reason
+      }
+    }
+
+    mutating func makeConnection(
+      to addresses: [SocketAddress],
+      using connector: any HTTP2Connector,
+      backoff: ConnectionBackoff,
+      defaultCompression: CompressionAlgorithm,
+      enabledCompression: CompressionAlgorithmSet
+    ) -> Connection? {
+      switch self {
+      case .notConnected:
+        var iterator = addresses.makeIterator()
+        let address = iterator.next()!  // addresses must not be empty.
+
+        let connection = Connection(
+          address: address,
+          http2Connector: connector,
+          defaultCompression: defaultCompression,
+          enabledCompression: enabledCompression
+        )
+
+        let connecting = State.Connecting(
+          connection: connection,
+          addresses: addresses,
+          addressIterator: iterator,
+          backoff: backoff.makeIterator()
+        )
+
+        self = .connecting(connecting)
+        return connection
+
+      case .connecting, .connected, .closing, .closed:
+        return nil
+      }
+    }
+
+    enum OnClose {
+      case none
+      case shutdown
+      case close(Connection)
+    }
+
+    mutating func close() -> OnClose {
+      let onClose: OnClose
+
+      switch self {
+      case .notConnected:
+        onClose = .shutdown
+
+      case .connecting(let state):
+        self = .closing(Closing(from: state, reason: .user))
+        // Do nothing; the connection hasn't been established yet so can't be closed.
+        onClose = .none
+
+      case .connected(let state):
+        self = .closing(Closing(from: state, reason: .user))
+        onClose = .close(state.connection)
+
+      case .closing(var state):
+        switch state.reason {
+        case .user:
+          onClose = .none
+        case .goAway:
+          // Override the reason: user closing takes precedence.
+          state.reason = .user
+          onClose = .close(state.connection)
+        }
+        self = .closing(state)
+
+      case .closed:
+        onClose = .none
+      }
+
+      return onClose
+    }
+
+    enum OnConnectSucceeded {
+      case updateState
+      case close(Connection)
+      case none
+    }
+
+    mutating func connectSucceeded() -> OnConnectSucceeded {
+      switch self {
+      case .connecting(let state):
+        self = .connected(Connected(from: state))
+        return .updateState
+      case .closing(let state):
+        self = .closing(state)
+        return .close(state.connection)
+      case .notConnected, .connected, .closed:
+        return .none
+      }
+    }
+
+    enum OnConnectFailed {
+      case none
+      case connect(Connection)
```

Comment:
I'm not sure because we're effectively telling the caller to call `connect()` on the `Connection` that we give them, I'm not sure that the caller needs to know it's the next address.